### PR TITLE
DiagnosticScope : Add SetDisplayName Function for an activity

### PR DESCRIPTION
**Pull request by sourabh1007**
Adding `SetDisplayName ` function for an activity created by diagnostic scope. and handle a scenario where it is not available.

### CODE REVIEW

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 1]
```diff
@@ -24,7 +24,8 @@ namespace Azure.Core.Pipeline
         private readonly ActivityAdapter? _activityAdapter;
         private readonly bool _suppressNestedClientActivities;
 
-        internal DiagnosticScope(string scopeName, DiagnosticListener source, object? diagnosticSourceArgs, object? activitySource, ActivityKind kind, bool suppressNestedClientActivities)
+        internal DiagnosticScope(string scopeName, DiagnosticListener source, object? diagnosticSourceArgs, object? activitySource, ActivityKind kind, bool suppressNestedClientActivities,
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:27]: @JoshLove-msft wdyt?

[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:27]: Yeah, agreed.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScopeFactory.cs [hunk 2]
```diff
@@ -45,7 +45,7 @@ public DiagnosticScopeFactory(string clientNamespace, string? resourceProviderNa
 
         public bool IsActivityEnabled { get; }
 
-        public DiagnosticScope CreateScope(string name, DiagnosticScope.ActivityKind kind = DiagnosticScope.ActivityKind.Internal)
+        public DiagnosticScope CreateScope(string name, DiagnosticScope.ActivityKind kind = DiagnosticScope.ActivityKind.Internal, string? displayName = null)
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScopeFactory.cs:48]: Agreed

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: added a setter.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScopeFactory.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScopeFactory.cs]: done

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 3]
```diff
@@ -566,6 +581,30 @@ public static bool GetIsAllDataRequested(this Activity activity)
             return GetAllDataRequestedMethod(activity);
         }
 
+        public static void SetDisplayName(this Activity activity, string? displayName)
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:584]: nit: 
```suggestion
        public static void SetDisplayName(this Activity activity, string displayName)
```

**Review by lmolkova — APPROVED** [review]
LGTM, just left on comment to review nullability

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:584]: Agreed this should not be nullable.

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 3 above)
Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:584]: looks like `ActivityAdaptor ` does not allow "non-nullable" value if not set in the constructor.

Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:584]: the parameter should not be nullable. The field can remain nullable.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 4]
```diff
@@ -92,6 +92,11 @@ public void Start()
             started?.SetCustomProperty(AzureSdkScopeLabel, AzureSdkScopeValue);
         }
 
+        public void SetDisplayName(string displayName)
```
Comment (kirankumarkolli) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:95]: Contract hints that the DisplayName is mutable right? 

//c: @lmolkova thoughts?

Comment (lmolkova) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:95]: it definitely is, event the underlying [DisplayName](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.displayname?view=net-7.0) property on the Activity is mutable.

**Pull request merged**